### ISSUE TITLE
Add Smarty to framework test system

### ISSUE THREAD

**Issue by MarkMaldaba** [facebook/hhvm#2763]
Hi there,

It would be great if you could add the Smarty templating engine (http://www.smarty.net/) to the list of frameworks that you test against (http://hhvm.com/frameworks/).

**Comment by uwetews** [comment]
I want to give you a status update. 3.1.22 will become a major release with a lot of improvements. So it takes some more time. Mit meanwhile I have done some testing with Smarty under HHVM.

The hackificator tool was able to convert most of the Smarty code automatically. 
With some work we should get Smarty itself 100% compatible.

The compiled template and cache PHP files are currently not compatible. But some brief tests have been successfull that we could generate these files directly as <?hh files if Smarty detects that it is running under HHVM. 

My goal is to have HHVM support included in the 3.1.22 release. 
Current estimated time frame 2 - 3 weeks.

I will keep you informed.

Uwe

**Comment by invacio** [comment]
What sort of work will need to be done locally to update smarty?

**Comment by fredemmott** [comment]
To close this task, it needs a phpunit test framework that's reliably included in the main repository.

If the current PHP code does not work, we will almost certainly consider it a bug in HHVM; no changes to smarty should be required for HHVM.

Converting to Hack could provide some developer-efficiency benefits, but that's really separate to this issue.

**Comment by invacio** [comment]
Hi there Fred, 

PhpUNIT test Framework? - You mean this is already included? 

Apologies HHVM is very new to me and most i discuss it with, so just to confirm.

We can add HHVM on to the current SmartyPHP version without any issues and it will run without any issues?

Converting to Hack -... apologies what is this in ref to?

**Comment by fredemmott** [comment]
- This issue is just for adding Smarty to http://hhvm.com/frameworks/
- hhvm.com/frameworks/ can run tests that use PHPUnit
- There are PHPUnit tests for Smarty, however, they're not currently in the source tree, so we can't use them
- See http://hacklang.org/ for 'converting to Hack'

Given we don't currently use Smarty ourselves, and the unit tests aren't currently available, I can't comment on if it will work or not.

**Comment by uwetews** [comment]
You can install the 3.1  releases with no problems under HHVM.

If you run the hackificator (see http://docs.hhvm.com/manual/en/install.hack.conversion.php) 
The majority of code gets converted to <?hh. Some modules will stay as PHP files.

The upcomming release 3.1.22 of Smarty will include optimizations for HHVM. The goal is that the hackificator can convert the complete Smarty distribution to <?hh.

3.1.22 will include also many other improvements resulting also under PHP in some higher speed and less memory usage.

I'm currently preparing a package including Smarty PHPUnit tests ready to run under HHVM. I will drop here a note when this work is completed.

**Comment by invacio** [comment]
Great, cant wait to see it released...

**Comment by invacio** [comment]
Q: Is this for finnished projects or can be used in test/dev framework also?

**Comment by uwetews** [comment]
Here is a status update.
Currently https://github.com/facebook/hhvm/issues/4797 does create problems in the PHPUnit test while testing Smarty's recompile and caching functions.

Smarty itself currently is affected when caching  mode CACHING_LIFETIME_SAVED is used.
(An update of an invalid cache file will be seen with a delay of one page request)

The current working copies can be required by composer with
       "smarty/smarty": "~3.1@dev"
        "smarty/smarty-phpunit": "~3.1@dev",

Run the test with 
$ sh phpunit_hhvm.sh

**Comment by invacio** [comment]
Hi there, Just to confirm is this pre-release issue likely to be resolved with the 3.22 release only im waiting on its release to update the entire site from 3.12 to 3.22 then add HHVM set up after through testing.

**Comment by uwetews** [comment]
The goal is to get this resolved with the 3.1.22 release.
But currently it looks like that we have to wait for a HHVM update to remove this PHP incompatibility.

**Comment by MarkMaldaba** [comment]
Isn't the point of this ticket just to get the Smarty unit tests into a public Git repo and add a submodule or whatever so they are included in the HHVM framework tests?

I don't think it really matters whether the tests pass or whether Smarty actually supports HHVM, does it?  The point is that the Smarty unit tests are being run under HHVM so that we get some transparency about compatibility (which, as has been said, is really a matter for the HHVM team to deal with, rather than the Smarty developer(s)).

Is there no public repo for Smarty that includes the unit tests?  There certainly used to be!

**Comment by fredemmott** [comment]
Yep, we're happy to start running the tests even if they don't all pass, once there is a canonical place for us to get them and the source from.

**Comment by uwetews** [comment]
The Smarty source and PHPUnit test can be installed via composer or are available at https://github.com/smarty-php/smarty and https://github.com/smarty-php/smarty-phpunit.

See my comment 3 days ago.

**Comment by fredemmott** [comment]
Cool; we need a unified repository - would it be possible for smarty-phpunit to require smarty via composer?

**Comment by uwetews** [comment]
I will add that.
So I will remove all @runInSeparateProcess annotation which I tried to find a work around until the include problem is solved. 
In the unit test we run tests on some demo custom cache resources like memcache, mysql, APC.
These may require some system configuration to run correctly. Shall I disable these test for the time being?

As it's already late night here in Germany I would update the smarty-phpunit tomorrow and let you know,

**Comment by invacio** [comment]
@fredemmott , @uwetews have we got an update on this yet and a possible lead time as to whether 3.22 will be launching 100% HM compat? 

or will 3.22 be launching with its generic updates and % compat ?

Thank you.

**Comment by billijaywest** [comment]
I second ^

**Comment by NotApplicableBeech** [comment]
Quote @uwetews 

I want to give you a status update. 3.1.22 will become a major release with a lot of improvements. So it takes some more time. Mit meanwhile I have done some testing with Smarty under HHVM.

The hackificator tool was able to convert most of the Smarty code automatically. 
With some work we should get Smarty itself 100% compatible.

The compiled template and cache PHP files are currently not compatible. But some brief tests have been successfull that we could generate these files directly as <?hh files if Smarty detects that it is running under HHVM.

My goal is to have HHVM support included in the 3.1.22 release. 
Current estimated time frame 2 - 3 weeks.

I will keep you informed.

Uwe


////

Have we got any update on this?

**Comment by NotApplicableBeech** [comment]
.......... any update chaps

**Comment by uwetews** [comment]
Sorry, the work on the 3.1.22 release takes longer as previously estimated.
There will be a couple of internal changes for less memory usage, higher speed and best performance under hhvm.

Also the PHPUnit test will be reworked to run under hhvm.

I now estimate another 1 - 2 weeks.

**Comment by NotApplicableBeech** [comment]
^ Looking forward to and happily awaiting...

**Comment by NotApplicableBeech** [comment]
@uwetews  Hi, any update on this release ? coming up to 3 weeks.

**Comment by MarkMaldaba** [comment]
Am I missing something?  All this ticket needs is a URL to the Git repository containing the unit tests.  We don't need to wait for a release for this, do we?

Even if 100% of tests fail under HHVM, that would be enough to satisfy this ticket!

**Comment by NotApplicableBeech** [comment]
@Mark we are awaiting Smarty 3.2 release and its 100% compatibility with HHVM

**Comment by MarkMaldaba** [comment]
@NotApplicableBeech - I am aware of that, and am looking forward to it with great anticipation!

However, that is not a requirement of this ticket.  This ticket is about linking the Smarty unit tests into the HHVM unit-testing framework, so we can get an idea of how well HHVM supports Smarty.  This will allow the HHVM developers to fix any bugs/edge-cases in HHVM that might cause Smarty to not work correctly.

There is no requirement for Smarty devs to make *any* change to Smarty in order for this to be possible, and given that there have been unit tests in Smarty since at least v2, this ticket should have been an easy fix.  The only thing holding it up is having a URL to the Git repository that contains the tests.

**Comment by uwetews** [comment]
Fred

 

Sorry for the delay in this matter.

 

I had for personal reason just little time to work on Smarty.

Also I had to make sure that our PHPUnit tests will fail on HHVM just because incompatibilities when executing Smarty, but not because of the test code.

 

You can now use composer  “smarty/smarty-dev” : “~3.1” to get the latest stable Smarty version and its corresponding PHPUnit tests.

 

Smarty-dev  contains just a wrapper composer.json  t load the versions from “smarty/smarty” and “smarty/smarty-phpunit”.

 

I hope this does work for you.

 

The failures you will get when running the unit test should all be caused by the reported issue https://github.com/facebook/hhvm/issues/4797#issuecomment-73614165

 

Besides that have not experience other problems.

 

The failure will not be a hard one in applications.

It will delay the visibility of a modified template source or an updated template cache by one page call.

 

Let me know if you need more information.

 

Uwe

 

Von: Fred Emmott [mailto:notifications@github.com] 
Gesendet: Dienstag, 10. Februar 2015 00:56
An: facebook/hhvm
Cc: uwetews
Betreff: Re: [hhvm] Add Smarty to framework test system (#2763)

 

Cool; we need a unified repository - would it be possible for smarty-phpunit to require smarty via composer?

**Comment by JoelMarcey** [comment]
There is an internal diff to add framework test runner support for Smarty.

**Comment by uwetews** [comment]
I'm glad to see that Smarty is now part of your frameworks tests.
But as far as I see you compare the test results with hphp/test/frameworks/results/smarty.expect. Right?

Currently a lot of development is under way at Smarty. So the PHPUnit tests will be extended regularly so I assume you might see no real failures.

Anyway  anyway I was able to implement a workaround for the existing HHVM issue https://github.com/facebook/hhvm/issues/4797#issuecomment-73614165

The dev-master version of Smarty is now passing all PHPUnit tests under HHVM.

I did set up at Travis CI for Smarty also a hhvm test run.

See https://travis-ci.org/smarty-php/smarty

So  every commit to the Smarty repository will now be automatically tested for hhvm compatibility.

Thanks for your good work.

Uwe
